src/main.cpp

```diff
@@ -2080,6 +2081,50 @@ void PartitionCheck(bool (*initialDownloadCheck)(), CCriticalSection& cs, const
     }
 }
 
+// Protected by cs_main
+static VersionBitsCache versionbitscache;
+
+int32_t ComputeBlockVersion(const CBlockIndex* pindexPrev, const Consensus::Params& params)
+{
+    LOCK(cs_main);
+    int32_t nVersion = 0;
+
+    for (int i = 0; i < (int)Consensus::MAX_VERSION_BITS_DEPLOYMENTS; i++) {
+        ThresholdState state = VersionBitsState(pindexPrev, params, (Consensus::DeploymentPos)i, versionbitscache);
+        if (state == THRESHOLD_LOCKED_IN || state == THRESHOLD_STARTED) {
+            nVersion |= VersionBitsMask(params, (Consensus::DeploymentPos)i);
+        }
+    }
+
+    if (nVersion == 0) {
+        // Pre-versionbits
+        return VERSIONBITS_LAST_OLD_BLOCK_VERSION;
+    } else {
+        return nVersion | VERSIONBITS_TOP_BITS;
+    }
+}
+
+/**
+ * Threshold condition checker that triggers when unknown versionbits are seen on the network.
+ */
+class WarningBitsConditionChecker : public AbstractThresholdConditionChecker
+{
+public:
+    int64_t BeginTime(const Consensus::Params& params) const { return 0; }
+    int64_t EndTime(const Consensus::Params& params) const { return std::numeric_limits<int64_t>::max(); }
+    int Period(const Consensus::Params& params) const { return params.nMinerConfirmationWindow; }
+    int Threshold(const Consensus::Params& params) const { return params.nRuleChangeActivationThreshold; }
+
+    bool Condition(const CBlockIndex* pindex, const Consensus::Params& params) const
+    {
+        return (pindex->nVersion & ~ComputeBlockVersion(pindex->pprev, params)) != 0;
```

Comment:
Won't this have potential false alerts, if let's say half the blocks in a period are signaling some unknown bit i, while the other half are signaling some other unknown bit j?

Anyway it occurred to me that it might be nice to record which bits are being set that are unknown and write that to the debug log when we warn a softfork is about to activate (as a useful debugging/investigation aid), but I can see that is a little tricky.